draft-holmberg-mmusic-ice-premature.xml

```diff
@@ -58,35 +58,33 @@
         for Network Address Translator (NAT) traversal for UDP-based communication.
       </t>
       <t>
-        When an endpoint receives the initial set of candidates from the remote peer, it starts to
-        process them. As part of the processing, one or more candidates might get discarded. There
-        are a number of reasons why a candidate is discared, e.g., because it represents an IP version
-        that is not supported by the endpoint, or if the connection-address associated with the candidate
-        is represented by a Fully Qualified Domain Name (FQDN) that the endpoint is not able to resolve.
-        Similairly, as a candidate by definition represents a single connection-address, if a FQDN
-        resoles into multiple connection-addresses the associated candidate needs to be discared.
+        Typically, when using ICE, endpoints will exchange candidate addresses,
+        form a list of candidate pairs, and then test each candidate pair to see
+        if connectivity can be established. If the test for a given pair fails,
+        it is marked accordingly, and if all pairs have failed, the overall
+        ICE process is considered to have failed.
       </t>
-      <t>
-        NOTE: Future specifications might describe how FQDNs that resolve into multiple connection-addresses
-        are used with ICE.
-      </t>
-      <t>
-        According to the procedures in <xref target="RFC8445"></xref>, if an endpoint discards the complete set
-        of initial candidates received from the remote peer, it will declare an ICE failure, and terminate
-        the session establishment, as the endpoint's list of remote candidates will be empty.
+      <t>During the process of connectivity checks, additional candidates may
+        be created as a result of successful inbound checks. Such candidates
+        are referred to as peer-reflexive candidates, and once discovered, will
+        be used to form new candidate pairs which will be tested like any other.
+        However, there is an inherent race condition here; if an endpoint
+        runs out of viable candidate pairs before learning about the
+        peer-reflexive candidates, it will prematurely fail the overall ICE
+        process. This race condition can occur in many common situations.
       </t>
```

Comment:
Q1: Instead of "fail the overall process" I would say something like "declare failure and terminate the ICE process".

Q2: Eventhough it's clear to people who know ICE, I think it would be good to clarify that peer-reflexive candidates will not be created until the remote peer starts sending connectivity checks.

Something like:

"...inherent race condition here; since peer-reflexive candidates will not be created before the remote peer begins performing its connectivity test checks, if an endpoint runs out of..."

Q3: Also, not sure whether the "This race condition can..." sentence is needed.

Reply:
Agree with Q1. 

I think Q3 is useful because it's not immediately obvious that this is a real-world problem.

Reply:
Ok, I will withdraw Q3 for now.

Reply:
Adopted Q1 and added some text to address Q2.